### ISSUE TITLE
[rl] Add `validate()` to training configs — on-policy (PPO/PPG)

### ISSUE THREAD

**Issue by anthonytorlucci** [anthonytorlucci/rlevo#166]
- **Labels:** bug
- **Files:** `algorithms/ppo/ppo_config.rs`, `algorithms/ppg/ppg_config.rs`
- **Body:** `PpoTrainingConfig::minibatch_size()` can return `0` (ppo §1.1, a genuine bug — needs `.max(1)` to match `update`); PPO has no `validate()`/`PpoConfigError` and `num_envs` should be removed since only `1` is valid (ppo §1.2/3.1/3.4). PPG lacks `validate()` entirely, permitting `beta_clone` non-finite/negative and `e_aux == 0`/`n_iteration == 0`, and the dead `InvalidConfig` error variant should be resurrected (ppg §5.1/1.1/1.2/1.3). Source: `-rl-ppo.md`, `-rl-ppg.md`.
- **Severity:** Contains High.

**Comment by anthonytorlucci** [comment]
Resolved by the ADR 0026 rollout (`9341350`).

`PpoTrainingConfig` and `PpgConfig` now `impl Validate` (PPG delegates `self.ppo.validate()?` then checks its own `n_iteration`/`e_aux`/`aux_batch_size > 0`, `beta_clone >= 0`). Builder `.build() -> Result`, agent `new -> Result`.

The two `assert_eq!(config.num_envs, 1)` panics in `ppo_agent.rs` / `ppg_agent.rs` are **deleted** and replaced by a `ConstraintKind::Custom("v1 supports num_envs == 1 only")` check in `PpoTrainingConfig::validate` — confirmed end-to-end: `PpoAgent::new` rejects `num_envs = 2`. Standard invariants (`num_steps/num_minibatches/update_epochs > 0`, `gamma/gae_lambda in [0,1]`, `clip_coef > 0`, `target_kl > 0` if `Some`) are covered.

**Comment by anthonytorlucci** [comment]
Re-verified against the current `crates/rlevo-reinforcement-learning/src/algorithms/ppo/ppo_config.rs` on `main`: the `validate()`/`build() -> Result`/`num_envs` fixes described in the closing comment are all confirmed live. **However, the §1.1 `minibatch_size()` zero-return bug this issue's body also names was never fixed** and is still present:

`ppo_config.rs:107-109`
```rust
pub fn minibatch_size(&self) -> usize {
    self.batch_size() / self.num_minibatches.max(1)
}
```

This still disagrees with the actual training loop's `mb_size`, `ppo_agent.rs:470`:
```rust
let mb_size = (batch_size / self.config.num_minibatches.max(1)).max(1);
```

Executed repro (`num_envs=1, num_steps=10, num_minibatches=20`):
```
public method minibatch_size() = 0
actual update() mb_size = 1
```

The public API and the training semantics still disagree — exactly the divergence this issue was filed to close. Reopening; the fix is the one-line `.max(1)` on the quotient from the original review (§1.1 of `code-review-30-06-2026-rl-ppo.md`).

**Comment by anthonytorlucci** [comment]
Update against current code: the PPG-config portion of this issue is **already fixed**, independent of any PR against this issue.

`crates/rlevo-reinforcement-learning/src/algorithms/ppg/ppg_config.rs:75-91` — `PpgConfig` implements `Validate` (`config::nonzero` for `n_iteration`/`e_aux`/`aux_batch_size`, `config::in_range` for `beta_clone`, delegating to `self.ppo.validate()`), and `PpgConfigBuilder::build()` (`:164-167`) returns `Result<PpgConfig, ConfigError>`, calling `validate()`. `PpgAgent::new` (`ppg_agent.rs:247-254`) calls `config.validate()?` and returns `Result<Self, ConfigError>` — no `assert_eq!` panic remains. Tests `default_config_is_valid`, `rejects_zero_e_aux`, `delegates_to_wrapped_ppo` (`ppg_config.rs:214-232`) exercise this.

So the review's §5.1/1.2/1.3 claims (against `-rl-ppg.md`) are refuted as currently filed. Two things genuinely remain, filed separately so they don't get lost under a "closed" label:
- `beta_clone` validation has an inclusive-infinity gap (`+inf` still passes `in_range(0.0, f64::INFINITY, ...)`) — filed as #502.
- `PpgAgentError::InvalidConfig(String)` (`ppg_agent.rs:57`) is still never constructed — `PpgAgent::new` returns the bare `rlevo_core::config::ConfigError`, not `PpgAgentError`, so the variant this issue asked to "resurrect" is still dead. Flagged as a sibling of #484 (comment added there).

The PPO-side items this issue also names (`minibatch_size()` can return 0, `num_envs` removal) are unrelated to `ppg_config.rs` and remain open in `ppo_config.rs` — not verified here.

**Comment by anthonytorlucci** [comment]
Verified against current \`main\`: this is resolved on both halves of the body.

- \`crates/rlevo-reinforcement-learning/src/algorithms/ppg/ppg_config.rs:75-90\`: \`impl Validate for PpgConfig\` now checks \`n_iteration\` nonzero, \`beta_clone\` finite/bounded, \`aux_batch_size\` nonzero, and delegates to \`self.ppo.validate()\`.
- \`crates/rlevo-reinforcement-learning/src/algorithms/ppg/ppg_agent.rs:247-254\`: \`PpgAgent::new\` returns \`Result<Self, ConfigError>\` and calls \`config.validate()?\` first thing — the old \`assert_eq!(num_envs, 1)\` panic is gone.
- \`crates/rlevo-reinforcement-learning/src/algorithms/ppo/ppo_config.rs:102-116,143-152\`: \`PpoTrainingConfig::minibatch_size()\` is \`self.batch_size() / self.num_minibatches.max(1)\` (the \`.max(1)\` guard from §1.1 is in place), and \`impl Validate for PpoTrainingConfig\` rejects \`num_envs != 1\`.

The dead \`PpgAgentError::InvalidConfig\` variant mentioned in the original body no longer applies — PPG's constructor now returns \`rlevo_core::config::ConfigError\`, not \`PpgAgentError\`, for validation failures (worth confirming that's the intended error-type story, but it's out of scope for this issue).

Recommend closing. Source: code review 2026-06-30 (\`code-review-30-06-2026-rl-ppg.md\` §3.1/3.2).
